src/backend/optimizer/prep/prepunion.c

```diff
@@ -1558,9 +1560,36 @@ expand_inherited_rtentry(PlannerInfo *root, RangeTblEntry *rte, Index rti)
 			root->rowMarks = lappend(root->rowMarks, newrc);
 		}
 
-		/* Close child relations, but keep locks */
 		if (childOID != parentOID)
-			heap_close(newrelation, NoLock);
+		{
+			LOCKMODE	releaseLockMode;
+
+			/*
+			 * GPDB: Decide whether we should relinquish locks on children.
+			 *
+			 * If gp_keep_partition_children_locks is on, and we have an AO child,
+			 * then we keep the lock. This is the preferred behavior for strict
+			 * correctness for concurrent AO vacuums.
+			 *
+			 * Otherwise, we defer to rel_is_part_child(), which taps into the
+			 * GP specific optimization of locking only the parent in a partition
+			 * hierarchy, to save on memory and lock consumption inside a tx.
+			 * For example, if we didn't release the locks here, and we were
+			 * doing a join between two partition roots, with say P1 and P2
+			 * number of children, we would have (P1 + P2) locks being held till
+			 * end of command. Similarly, in the same transaction if we operate
+			 * on these 2 partition hierarchies in any way, we would be holding
+			 * (P1 + P2) locks till the end of the transaction.
+			 */
+			if (RelationIsAppendOptimized(newrelation) &&
+				(parent_is_partitioned || parent_is_interior_partition) &&
+				gp_keep_partition_children_locks)
+				releaseLockMode = NoLock;
+			else
+				releaseLockMode = rel_is_part_child(childOID) ? NoLock: lockmode;
```

Comment:
Let's preserve the behaviour of previous patch and keep holding locks as in the patch.

Reply:
@bandetto, let's bring back the DDL's code. I'll try to introduce the dependancy from GUC in separate path later.